### ISSUE TITLE
[java] XPath type attribute available besides typeIs function

### ISSUE THREAD

**Issue by jborgers** [pmd/pmd#5690]
**Is your feature request related to a problem? Please describe.**
typeIs() is very useful to determine the type of a node. However, it is limited to comparing one node with a literal. Sometimes you want to compare types of two nodes, or types of two list of nodes, which is not possible now.

**Describe the solution you'd like**
We'd like to have more powerful support for comparing types of nodes.

**Describe alternatives you've considered**
Use of typeIs() is too limited.

**Examples**
We created a rule AvoidInfiniteRecursion, however it was really complicated to write one without being able to compare types of the formal parameters and the method call arguments, and it has false positives.

With additional attributes this would be much more easy.
Proposed new attributes: @ Type and @ MemberOfType 
Example XPath: 
``xpath
//MethodDeclaration
(: call same method :)
//MethodCall[@MethodName = ancestor::MethodDeclaration/@Name and @MemberOfType = ancestor::ClassDeclaration[1]/@Type
    and deep-equal(ArgumentList/VariableAccess/@Type = ancestor::MethodDeclaration/FormalParameters/FormalParameter/@Type)]
``
Difficulty to consider: the difference between `typeIs` and `typeIsExactly`. @ Type attribute should be the exact type (maybe named @ ExactType), and typeIs(node/@ Type) could be used to include super types. So simplified like:
`ArgumentList/VariableAccess[pmd-java:typeIs(FormalParameter/@Type)]`
or maybe even better would be including the deep-equals, something like:
`pmd-java:typesAre(ArgumentList/VariableAccess, ancestor::MethodDeclaration/FormalParameters/FormalParameter)`

**Comment by jsotuyod** [comment]
I don't think this would work because of things such as type coercion (ie: boxing and unboxing).

The most realistic approach here would be to use `pmd-java:matchesSig` to match a `MethodCall` vs a method signature. The one thing missing however right now I believe is that we don't expose method signatures for method declarations, so right now all usages are limited to a static method signature in the rule / properties.

**Comment by jborgers** [comment]
Right, being able to compare the method signature for a MethodCall and a MethodDeclaration would do it! 
Doesn't type coercion be complicating this? I don't see if it does for dynamic type comparison, for what reason not for dynamic method signature comparison.

**Comment by oowekyala** [comment]
I think this is a case where the Java API already allows you to do this, and it would be a better idea to implement the rule in Java. XPath has problems with scalability. Attributes may only be strings or primitives, so you would have to give up all type safety to use attributes. Providing XPath functions allow us to circumvent this partly, but they still have problems because XPath expressions have sequence types and therefore need to be iterated on, either within the function or outside of it in XPath. So it's only reasonable to provide simple functions that have well defined semantics.

For instance a `matchesSig` XPath function that accepts a MethodCall and a method signature would have to ask itself, is it valid to pass several MethodCall nodes to the first argument and how does it behave then, where does the method signature come from (an attribute? a function?), what type does it have (likely must be a string), how to parse it, etc. If you write your rule in Java you have access to a very rich API to represent symbols and types without mapping them to strings and back. The control flow is clearer and you avoid a lot of inefficiencies. And if you want to make your rule more complex later it's going to be much easier in Java than in XPath.

```ruby
//MethodDeclaration
(: call same method :)
//MethodCall[@MethodName = ancestor::MethodDeclaration/@Name and @MemberOfType = ancestor::ClassDeclaration[1]/@Type
    and deep-equal(ArgumentList/VariableAccess/@Type = ancestor::MethodDeclaration/FormalParameters/FormalParameter/@Type)]
```
In the Java API you can just refer to the variable for the MethodCall instead of having to fetch it from the ancestor stream and duplicating work. The nested descendant axis `//MethodCall` is also looking too far, because it may enter anonymous classes and lambdas, which `descendants(ASTMethodCall.class)` doesn't do. A java rule will likely execute orders of magnitude faster. And you can use function from our utility classes for instance. 
```java

    @Override
    public Object visit(ASTMethodDeclaration node, Object data) {

        NodeStream<ASTMethodCall> recursiveCalls =
                node.descendants(ASTMethodCall.class)
                        .filter(it -> node.getName().equals(it.getMethodName()))
                        .filter(it -> node.getSymbol().equals(it.getMethodType().getSymbol()))
                        .filter(it -> isCallWithJustFormalParams(node, it));

        for (ASTMethodCall call : recursiveCalls) {
            asCtx(data).addViolation(call);
        }
        return data;
    }

    private static boolean isCallWithJustFormalParams(ASTMethodDeclaration node, ASTMethodCall it) {
        if (it.getArguments().size() != node.getFormalParameters().size()) {
            return false;
        }
        List<ASTVariableId> methodParams =
                node.getFormalParameters().toStream().toList(ASTFormalParameter::getVarId);
        for (int i = 0; i < methodParams.size(); i++) {
            ASTExpression arg = it.getArguments().get(i);
            ASTVariableId param = methodParams.get(i);
            if (!JavaAstUtils.isReferenceToVar(arg, param.getSymbol())) {
                return false;
            }
        }
        return true;
    }
```

So I think XPath rules are nice if you want to match a very simple pattern, but Java is more appropriate whenever you want to go any further.

**Comment by jborgers** [comment]
Okay, thanks for explaining @oowekyala, I see your point. 

That said, I do like the XPath way. One advantage is that custom rule files can easily be distributed, e.g. just by URL, without worries about security. With the improvements in pmd7 it has become much more expressive, too. 
With a more complex rule, the expression gets more difficult to understand and maintain, and speed and limitations are indeed also down sides, agreed. However, if complexity can be reduced with a few extra, powerful functions, I would be a happy consumer of those.
